mypy/binder.py

```diff
@@ -227,7 +227,11 @@ def assign_type(self, expr: Expression,
         if (isinstance(self.most_recent_enclosing_type(expr, type), AnyType)
                 and not restrict_any):
             pass
-        elif isinstance(type, AnyType):
+        elif (isinstance(type, AnyType)
+              and (not isinstance(declared_type, UnionType)
+                   or not any(isinstance(item, NoneTyp) for item in declared_type.items))):
+            # Assigning an Any value doesn't affect the type to avoid false negatives, unless
+            # there is an Any item in a declared union type.
```

Comment:
Do you mean "unless there is a _None_ item in a declared union type"? (because you are checking for ``NoneTyp``)